Table.openBlock fails on Windows 7 but works on Debian

The following code (using release 0.11):
```
   public ChainstateIterator() {
        Options options = new Options();
        options.createIfMissing(false);
        options.compressionType(CompressionType.NONE);
        String datadir = treasure.getDataDir();

        try {
            db = factory.open(new File(datadir + "chainstate" + File.separator), options);
            obfuscationKey = extractKey(db);
            System.out.println(new Bytes(obfuscationKey));
            levelDBIterator = db.iterator();
            levelDBIterator.seekToFirst();
        } catch (IOException e) {
            ...
        }
    }
```
works perfectly on Debian Stretch, amd64 but fails with:
```
Exception in thread "main" java.lang.RuntimeException: java.nio.channels.ClosedChannelException
	at com.google.common.base.Throwables.propagate(Throwables.java:241)
	at org.iq80.leveldb.table.Table.openBlock(Table.java:83)
	at org.iq80.leveldb.util.TableIterator.getNextBlock(TableIterator.java:102)
	at org.iq80.leveldb.util.TableIterator.seekInternal(TableIterator.java:57)
	at org.iq80.leveldb.util.TableIterator.seekInternal(TableIterator.java:26)
	at org.iq80.leveldb.util.AbstractSeekingIterator.seek(AbstractSeekingIterator.java:41)
	at org.iq80.leveldb.util.InternalTableIterator.seekInternal(InternalTableIterator.java:45)
	at org.iq80.leveldb.util.InternalTableIterator.seekInternal(InternalTableIterator.java:25)
	at org.iq80.leveldb.util.AbstractSeekingIterator.seek(AbstractSeekingIterator.java:41)
	at org.iq80.leveldb.impl.Level.get(Level.java:134)
	at org.iq80.leveldb.impl.Version.get(Version.java:172)
	at org.iq80.leveldb.impl.VersionSet.get(VersionSet.java:223)
	at org.iq80.leveldb.impl.DbImpl.get(DbImpl.java:616)
	at org.iq80.leveldb.impl.DbImpl.get(DbImpl.java:577)
	at nxt.crosschain.db.ChainstateIterator.extractKey(ChainstateIterator.java:110)
	at nxt.crosschain.db.ChainstateIterator.<init>(ChainstateIterator.java:44)
	at nxt.crosschain.db.LevelDBImportWindowsTest.main(LevelDBImportWindowsTest.java:67)
Caused by: java.nio.channels.ClosedChannelException
	at sun.nio.ch.FileChannelImpl.ensureOpen(FileChannelImpl.java:110)
	at sun.nio.ch.FileChannelImpl.read(FileChannelImpl.java:721)
	at org.iq80.leveldb.table.FileChannelTable.read(FileChannelTable.java:96)
	at org.iq80.leveldb.table.FileChannelTable.readBlock(FileChannelTable.java:55)
	at org.iq80.leveldb.table.Table.openBlock(Table.java:80)
	... 15 more
```
on Windows 7, x86 using JDK 1.8.0_161
what could be the problem with nio.channels?

Hi,
I'm getting similar exceptions on my Raspberry Pi. (My code works fine on Ubuntu and Windows.)
`Caused by: java.lang.RuntimeException: java.nio.channels.ClosedChannelException
        at com.google.common.base.Throwables.propagate(Throwables.java:240)
        at org.iq80.leveldb.table.Table.openBlock(Table.java:83)
        at org.iq80.leveldb.util.TableIterator.getNextBlock(TableIterator.java:1                                         02)
        at org.iq80.leveldb.util.TableIterator.getNextElement(TableIterator.java                                         :79)
        at org.iq80.leveldb.util.AbstractSeekingIterator.hasNext(AbstractSeeking                                         Iterator.java:48)
        at org.iq80.leveldb.util.InternalTableIterator.getNextElement(InternalTa                                         bleIterator.java:51)
        at org.iq80.leveldb.util.AbstractSeekingIterator.hasNext(AbstractSeeking                                         Iterator.java:48)
        at org.iq80.leveldb.util.DbIterator.resetPriorityQueue(DbIterator.java:1                                         55)
        at org.iq80.leveldb.util.DbIterator.<init>(DbIterator.java:75)
        at org.iq80.leveldb.impl.DbImpl.internalIterator(DbImpl.java:778)
        at org.iq80.leveldb.impl.DbImpl.iterator(DbImpl.java:744)
        at org.iq80.leveldb.impl.DbImpl.iterator(DbImpl.java:735)
        at org.iq80.leveldb.impl.DbImpl.iterator(DbImpl.java:84)
        at com.hunelco.forint.database.LevelDBPeerStore.getAll(LevelDBPeerStore.                                         java:127)
        ... 2 more
Caused by: java.nio.channels.ClosedChannelException
        at sun.nio.ch.FileChannelImpl.ensureOpen(FileChannelImpl.java:110)
        at sun.nio.ch.FileChannelImpl.read(FileChannelImpl.java:720)
        at org.iq80.leveldb.table.FileChannelTable.read(FileChannelTable.java:96                                         )
        at org.iq80.leveldb.table.FileChannelTable.readBlock(FileChannelTable.ja                                         va:55)
        at org.iq80.leveldb.table.Table.openBlock(Table.java:80)
        ... 14 more
`

> The following code (using release 0.11):
> 
> ```
>    public ChainstateIterator() {
>         Options options = new Options();
>         options.createIfMissing(false);
>         options.compressionType(CompressionType.NONE);
>         String datadir = treasure.getDataDir();
> 
>         try {
>             db = factory.open(new File(datadir + "chainstate" + File.separator), options);
>             obfuscationKey = extractKey(db);
>             System.out.println(new Bytes(obfuscationKey));
>             levelDBIterator = db.iterator();
>             levelDBIterator.seekToFirst();
>         } catch (IOException e) {
>             ...
>         }
>     }
> ```
> works perfectly on Debian Stretch, amd64 but fails with:
> 
> ```
> Exception in thread "main" java.lang.RuntimeException: java.nio.channels.ClosedChannelException
> 	at com.google.common.base.Throwables.propagate(Throwables.java:241)
> 	at org.iq80.leveldb.table.Table.openBlock(Table.java:83)
> 	at org.iq80.leveldb.util.TableIterator.getNextBlock(TableIterator.java:102)
> 	at org.iq80.leveldb.util.TableIterator.seekInternal(TableIterator.java:57)
> 	at org.iq80.leveldb.util.TableIterator.seekInternal(TableIterator.java:26)
> 	at org.iq80.leveldb.util.AbstractSeekingIterator.seek(AbstractSeekingIterator.java:41)
> 	at org.iq80.leveldb.util.InternalTableIterator.seekInternal(InternalTableIterator.java:45)
> 	at org.iq80.leveldb.util.InternalTableIterator.seekInternal(InternalTableIterator.java:25)
> 	at org.iq80.leveldb.util.AbstractSeekingIterator.seek(AbstractSeekingIterator.java:41)
> 	at org.iq80.leveldb.impl.Level.get(Level.java:134)
> 	at org.iq80.leveldb.impl.Version.get(Version.java:172)
> 	at org.iq80.leveldb.impl.VersionSet.get(VersionSet.java:223)
> 	at org.iq80.leveldb.impl.DbImpl.get(DbImpl.java:616)
> 	at org.iq80.leveldb.impl.DbImpl.get(DbImpl.java:577)
> 	at nxt.crosschain.db.ChainstateIterator.extractKey(ChainstateIterator.java:110)
> 	at nxt.crosschain.db.ChainstateIterator.<init>(ChainstateIterator.java:44)
> 	at nxt.crosschain.db.LevelDBImportWindowsTest.main(LevelDBImportWindowsTest.java:67)
> Caused by: java.nio.channels.ClosedChannelException
> 	at sun.nio.ch.FileChannelImpl.ensureOpen(FileChannelImpl.java:110)
> 	at sun.nio.ch.FileChannelImpl.read(FileChannelImpl.java:721)
> 	at org.iq80.leveldb.table.FileChannelTable.read(FileChannelTable.java:96)
> 	at org.iq80.leveldb.table.FileChannelTable.readBlock(FileChannelTable.java:55)
> 	at org.iq80.leveldb.table.Table.openBlock(Table.java:80)
> 	... 15 more
> ```
> on Windows 7, x86 using JDK 1.8.0_161
> what could be the problem with nio.channels?

same problem

--- a/leveldb/src/main/java/org/iq80/leveldb/impl/TableCache.java
+++ b/leveldb/src/main/java/org/iq80/leveldb/impl/TableCache.java
@@ -127,6 +127,8 @@ public class TableCache
                 FileChannel fileChannel = fis.getChannel();
                 if (Iq80DBFactory.USE_MMAP) {
                     table = new MMapTable(tableFile.getAbsolutePath(), fileChannel, userComparator, verifyChecksums);
+                    // We can close the channel and input stream as the mapping does not need them
+                    Closeables.closeQuietly(fis);
                 }
                 else {
                     table = new FileChannelTable(tableFile.getAbsolutePath(), fileChannel, userComparator, verifyChecksums);

--- a/leveldb/src/main/java/org/iq80/leveldb/impl/TableCache.java
+++ b/leveldb/src/main/java/org/iq80/leveldb/impl/TableCache.java
@@ -26,6 +26,7 @@ import org.iq80.leveldb.table.FileChannelTable;
 import org.iq80.leveldb.table.MMapTable;
 import org.iq80.leveldb.table.Table;
 import org.iq80.leveldb.table.UserComparator;
+import org.iq80.leveldb.util.Closeables;
 import org.iq80.leveldb.util.Finalizer;
 import org.iq80.leveldb.util.InternalTableIterator;
 import org.iq80.leveldb.util.Slice;
@@ -120,14 +121,20 @@ public class TableCache
         {
             String tableFileName = Filename.tableFileName(fileNumber);
             File tableFile = new File(databaseDir, tableFileName);
-            try (FileInputStream fis = new FileInputStream(tableFile);
-                    FileChannel fileChannel = fis.getChannel()) {
+            FileInputStream fis = null;
+            try {
+                fis = new FileInputStream(tableFile);
+                FileChannel fileChannel = fis.getChannel();
                 if (Iq80DBFactory.USE_MMAP) {
                     table = new MMapTable(tableFile.getAbsolutePath(), fileChannel, userComparator, verifyChecksums);
                 }
                 else {
                     table = new FileChannelTable(tableFile.getAbsolutePath(), fileChannel, userComparator, verifyChecksums);
                 }
+            }
+            catch (IOException ioe) {
+              Closeables.closeQuietly(fis);
+              throw ioe;
             }
         }
 

